File: ios/Sources/RouteMatch.swift
// RouteMatch — Swift port of src/live/route-match.ts (parity law, 2026-09-06).
// "Your history is your route": the phone carries every past GPS track; the
// live run's first few hundred meters identify which one he is on (either
// direction), and from then on the route AHEAD is known. Pure, no network.
import Foundation

struct RoutePoint {
  let lat: Double
  let lon: Double
  let distM: Double
  let altM: Double
}

struct Route {
  let id: String
  let km: Double
  /// How many times this route was run — the habit prior for ties.
  let runs: Int
  let points: [RoutePoint]
}

struct Fix {
  let lat: Double
  let lon: Double
  /// Live odometer, meters (watch distance — authoritative).
  let distM: Double
}

/// A terrain cue projected onto the LIVE run. Mirrors TS AheadCue.
struct AheadCue {
  let type: TerrainCue.Kind
  let liveDistanceM: Double
  let flatEquivRemainingM: Double
  let gainM: Double
  let confidence: Double
  let key: String
}

struct RouteLock {
  let routeId: String
  let reversed: Bool
  let progressM: Double
  let remainingM: Double
  /// Share of lock-eligible candidates agreeing on the position 300m ahead.
  let agreement: Double
}

/// The library file the relay serves: points as [lat, lon, distM, altM].
struct RouteLibraryFile: Decodable {
  struct R: Decodable {
    let id: String
    let km: Double
    let runs: Int
    let points: [[Double]]
  }
  let routes: [R]

  var asRoutes: [Route] {
    routes.map { r in
      Route(id: r.id, km: r.km, runs: r.runs,
            points: r.points.compactMap { p in p.count >= 4 ? RoutePoint(lat: p[0], lon: p[1], distM: p[2], altM: p[3]) : nil })
    }
  }
}

final class RouteMatcher {
  /// Mirrors TS constants (GUESS — the route backtest calibrates).
  static let lockMinTrackM = 500.0
  static let nearM = 30.0
  private static let lostM = 60.0
  private static let lostTravelM = 100.0
  private static let forwardWindowM = 150.0
  private static let backWindowM = 40.0
  private static let gridCellM = 60.0
  private let ky = 110_540.0

  private final class Candidate {
    let route: Route
    let reversed: Bool
    let key: String
    let sm: [ProfilePoint]
    let cumFlatEquivM: [Double]
    let cues: [TerrainCue]
    let grid: [String: [Int]]
    let kx: Double
    var started = false
    var idx = 0
    var progressM = 0.0
    var onRouteM = 0.0
    var offRouteM = 0.0
    init(route: Route, reversed: Bool, key: String, sm: [ProfilePoint], cum: [Double], cues: [TerrainCue], grid: [String: [Int]], kx: Double) {
      self.route = route; self.reversed = reversed; self.key = key; self.sm = sm
      self.cumFlatEquivM = cum; self.cues = cues; self.grid = grid; self.kx = kx
    }
  }

  private var cands: [Candidate] = []
  private var current: Candidate?
  private var lastFix: Fix?

  init(routes: [Route]) {
    for r in routes where r.points.count >= 10 {
      cands.append(candidate(r, reversed: false))
      cands.append(candidate(r, reversed: true))
    }
  }

  private func candidate(_ route: Route, reversed: Bool) -> Candidate {
    let total = route.points[route.points.count - 1].distM
    let pts: [RoutePoint] = reversed
      ? route.points.reversed().map { RoutePoint(lat: $0.lat, lon: $0.lon, distM: total - $0.distM, altM: $0.altM) }
      : route.points
    let r = Route(id: route.id, km: route.km, runs: route.runs, points: pts)
    let profile = pts.map { ProfilePoint(distanceM: $0.distM, altitudeM: $0.altM) }
    let sm = Terrain.smoothProfile(profile)
    var cum = [Double](repeating: 0, count: sm.count)
    if sm.count > 1 {
      for i in 1..<sm.count {
        let dd = sm[i].distanceM - sm[i - 1].distanceM
        if dd <= 0 { cum[i] = cum[i - 1]; continue }
        let grade = (sm[i].altitudeM - sm[i - 1].altitudeM) / dd
        let mult = Terrain.gapMultiplier(grade)
        cum[i] = cum[i - 1] + dd * mult
      }
    }
    let kx = 111_320 * cos(pts[0].lat * .pi / 180)
    var grid: [String: [Int]] = [:]
    for (i, p) in pts.enumerated() { grid[cell(kx, p.lat, p.lon), default: []].append(i) }
    return Candidate(route: r, reversed: reversed, key: "\(route.id)\(reversed ? ":rev" : "")",
                     sm: sm, cum: cum, cues: Terrain.extractTerrainCues(profile), grid: grid, kx: kx)
  }

  private func cell(_ kx: Double, _ lat: Double, _ lon: Double) -> String {
    "\(Int(floor(lon * kx / Self.gridCellM))),\(Int(floor(lat * ky / Self.gridCellM)))"
  }

  private func meters(_ kx: Double, _ aLat: Double, _ aLon: Double, _ bLat: Double, _ bLon: Double) -> Double {
    let dx = (aLon - bLon) * kx
    let dy = (aLat - bLat) * ky
    return (dx * dx + dy * dy).squareRoot()
  }

  /// Nearest route point (forward window when started, direction-aware grid
  /// search otherwise) and the runner's CONTINUOUS progress by projection
  /// onto the segment through it. Mirrors TS `nearest`.
  private func nearest(_ c: Candidate, _ fix: Fix, heading: (Double, Double)?) -> (Int, Double, Double) {
    let pts = c.route.points
    var best = -1
    var bestD = Double.infinity
    if c.started {
      let from = c.progressM - Self.backWindowM
      let to = c.progressM + Self.forwardWindowM
      var i = max(0, c.idx - 3)
      while i < pts.count && pts[i].distM <= to {
        if pts[i].distM >= from {
          let d = meters(c.kx, fix.lat, fix.lon, pts[i].lat, pts[i].lon)
          if d < bestD { bestD = d; best = i }
        }
        i += 1
      }
    } else {
      let cx = Int(floor(fix.lon * c.kx / Self.gridCellM))
      let cy = Int(floor(fix.lat * ky / Self.gridCellM))
      for dx in -1...1 {
        for dy in -1...1 {
          guard let arr = c.grid["\(cx + dx),\(cy + dy)"] else { continue }
          for i in arr {
            let d = meters(c.kx, fix.lat, fix.lon, pts[i].lat, pts[i].lon)
            if d > Self.nearM || d >= bestD { continue }
            if let h = heading, i + 1 < pts.count {
              let rx = (pts[i + 1].lon - pts[i].lon) * c.kx
              let ry = (pts[i + 1].lat - pts[i].lat) * ky
              if rx * h.0 + ry * h.1 <= 0 { continue }
            }
            bestD = d
            best = i
          }
        }
      }
    }
    if best < 0 { return (-1, .infinity, c.progressM) }
    var progress = pts[best].distM
    var residual = bestD
    for (a, b) in [(best - 1, best), (best, best + 1)] {
      if a < 0 || b >= pts.count { continue }
      let ax = pts[a].lon * c.kx, ay = pts[a].lat * ky
      let bx = pts[b].lon * c.kx, by = pts[b].lat * ky
      let px = fix.lon * c.kx, py = fix.lat * ky
      let vx = bx - ax, vy = by - ay
      let len2 = vx * vx + vy * vy
      if len2 <= 0 { continue }
      let tt = max(0, min(1, ((px - ax) * vx + (py - ay) * vy) / len2))
      let qx = ax + tt * vx, qy = ay + tt * vy
      let d = ((px - qx) * (px - qx) + (py - qy) * (py - qy)).squareRoot()
      if d < residual {
        residual = d
        progress = pts[a].distM + tt * (pts[b].distM - pts[a].distM)
      }
    }
    return (best, residual, progress)
  }

  /// Feed one GPS fix (≈1Hz). Mirrors TS `update`.
  func update(_ fix: Fix) {
    let dLive = lastFix.map { max(0, fix.distM - $0.distM) } ?? 0
    var heading: (Double, Double)?
    if let last = lastFix {
      let kx = 111_320 * cos(fix.lat * .pi / 180)
      let hx = (fix.lon - last.lon) * kx
      let hy = (fix.lat - last.lat) * ky
      if (hx * hx + hy * hy).squareRoot() >= 3 { heading = (hx, hy) }
    }
    lastFix = fix
    for c in cands {
      let (i, res, prog) = nearest(c, fix, heading: heading)
      if !c.started {
        if i >= 0 && res <= Self.nearM {
          c.started = true
          c.idx = i
          c.progressM = prog
          c.onRouteM = 0
          c.offRouteM = 0
        }
        continue
      }
      if i >= 0 && res <= Self.nearM {
        if prog >= c.progressM - Self.backWindowM {
          c.idx = i
          c.progressM = max(c.progressM, prog)
        }
        c.onRouteM += dLive
        c.offRouteM = 0
      } else if i < 0 || res > Self.lostM {
        c.offRouteM += dLive
        if c.offRouteM >= Self.lostTravelM {
          c.started = false
          c.onRouteM = 0
          c.offRouteM = 0
          if current === c { current = nil }
        }
      }
    }
    var best: Candidate?
    for c in cands where c.started && c.onRouteM >= Self.lockMinTrackM {
      if best == nil || better(c, best!) { best = c }
    }
    if let cur = current, cur.started, cur.onRouteM >= Self.lockMinTrackM {
      if let b = best, b !== cur, b.onRouteM > cur.onRouteM + Self.lockMinTrackM { current = b }
    } else {
      current = best
    }
  }

  private func better(_ a: Candidate, _ b: Candidate) -> Bool {
    if a.onRouteM != b.onRouteM { return a.onRouteM > b.onRouteM }
    if a.route.runs != b.route.runs { return a.route.runs > b.route.runs }
    return a.route.km > b.route.km
  }

  var lock: RouteLock? {
    guard let c = current else { return nil }
    let total = c.route.points[c.route.points.count - 1].distM
    return RouteLock(routeId: c.route.id, reversed: c.reversed, progressM: c.progressM, remainingM: total - c.progressM, agreement: agreement())
  }

  private func pointAhead(_ c: Candidate, _ aheadM: Double) -> RoutePoint? {
    let target = c.progressM + aheadM
    let pts = c.route.points
    var i = c.idx
    while i < pts.count {
      if pts[i].distM >= target { return pts[i] }
      i += 1
    }
    return nil
  }

  private func agreement(aheadM: Double = 300) -> Double {
    guard let cur = current else { return 0 }
    guard let mine = pointAhead(cur, aheadM) else { return 1 }
    var eligible = 0
    var agree = 0
    for c in cands where c.started && c.onRouteM >= Self.lockMinTrackM {
      eligible += 1
      if let p = pointAhead(c, aheadM), meters(cur.kx, mine.lat, mine.lon, p.lat, p.lon) <= 40 { agree += 1 }
    }
    return eligible > 0 ? Double(agree) / Double(eligible) : 1
  }

  /// Terrain cues AHEAD on the locked route, projected onto the live odometer.
  func aheadCues() -> [AheadCue] {
    guard let c = current, let fix = lastFix else { return [] }
    let here = flatEquivAt(c, c.progressM)
    let agreement = agreement()
    var out: [AheadCue] = []
    for cue in c.cues where cue.distanceM > c.progressM {
      out.append(AheadCue(
        type: cue.type,
        liveDistanceM: fix.distM + (cue.distanceM - c.progressM),
        flatEquivRemainingM: flatEquivAt(c, cue.distanceM) - here,
        gainM: cue.gainM,
        confidence: cue.confidence * (agreement >= 1 ? 1 : 0.5),
        key: "\(c.key)@\(Int(cue.distanceM.rounded()))"))
    }
    return out
  }

  /// The route's elevation ahead in LIVE-odometer space.
  func aheadProfile(maxM: Double = 2000) -> [ProfilePoint] {
    guard let c = current, let fix = lastFix else { return [] }
    var out: [ProfilePoint] = []
    for p in c.sm {
      if p.distanceM < c.progressM { continue }
      if p.distanceM > c.progressM + maxM { break }
      out.append(ProfilePoint(distanceM: fix.distM + (p.distanceM - c.progressM), altitudeM: p.altitudeM))
    }
    return out
  }

  private func flatEquivAt(_ c: Candidate, _ d: Double) -> Double {
    let sm = c.sm
    let cum = c.cumFlatEquivM
    var lo = 0
    var hi = sm.count - 1
    while lo < hi {
      let mid = (lo + hi) >> 1
      if sm[mid].distanceM < d { lo = mid + 1 } else { hi = mid }
    }
    if lo == 0 { return 0 }
    let a = sm[lo - 1]
    let b = sm[lo]
    let f = min(1, max(0, (d - a.distanceM) / max(1, b.distanceM - a.distanceM)))
    return cum[lo - 1] + f * (cum[lo] - cum[lo - 1])
  }
}

File: src/live/route-match.ts
// Route matching — "your history is your route" (design doc
// route-story-invisible, Approach B, live wiring 2026-09-06).
//
// The runner never downloads a route. The phone carries every past GPS
// track; the live run's first few hundred meters identify which one he is
// on (in either direction), and from then on the route AHEAD is known:
// its elevation profile feeds the terrain cues the replay lab proved
// (crest arrival p90 3.1s, drift-corrected). Diverge from the route and
// the lock drops within ~100m; another candidate may take over seamlessly.
//
// Pure module, ported to Swift same-session (parity law). No network.
import { extractTerrainCues, gapMultiplier, smoothProfile, type ProfilePoint, type TerrainCue } from './terrain'

export interface RoutePoint {
  lat: number
  lon: number
  /** Distance along the route, meters. */
  distM: number
  altM: number
}

export interface Route {
  id: string
  km: number
  /** How many times this route was run — the habit prior for ties. */
  runs: number
  points: RoutePoint[]
  /** Per point: share of all runs through here that were still on this
   *  route 100m later (history as a Markov chain, precomputed by the
   *  library builder). Absent = unknown = 1. */
  branch?: number[]
}

export interface Fix {
  lat: number
  lon: number
  /** Live odometer, meters (watch distance — authoritative). */
  distM: number
  /** Live altitude (barometric preferred) — lets the live track itself
   *  become a route with a profile (self-route). */
  altM?: number
}

/** A terrain cue projected onto the LIVE run. */
export interface AheadCue extends TerrainCue {
  /** Where the cue falls on the live odometer, meters. */
  liveDistanceM: number
  /** Flat-equivalent meters from the runner to the cue (GAP-weighted) —
   *  multiply by flat pace for the ETA. */
  flatEquivRemainingM: number
  /** Stable identity: route + direction + route distance. */
  key: string
  /** Probability (from history) the runner follows the route to this cue. */
  pReach: number
}

/** Live meters on-route before a candidate may be declared locked. Route
 *  backtest 2026-09-06 (300 runs, chronological leave-one-out): 500m →
 *  coverage 68.9% / wrong-future@50m 18.2%; 300m → 73.9% / 19.1%. 400m
 *  splits the difference; the same-front-door prior (1.5×) locks a
 *  familiar start in ~270m. */
export const LOCK_MIN_TRACK_M = 400
/** Residual (meters to the nearest route point) that counts as on-route.
 *  Phone GPS in a pocket is ~10m; opposite sidewalks are ~15-25m. */
export const NEAR_M = 30
/** Residual beyond which the candidate is being left... */
const LOST_M = 60
/** ...if it persists for this much live travel. */
const LOST_TRAVEL_M = 100
/** Forward search window for the progress pointer, meters. */
const FORWARD_WINDOW_M = 150
const BACK_WINDOW_M = 40

interface Candidate {
  route: Route
  reversed: boolean
  key: string
  /** Precomputed per route×direction. */
  sm: ProfilePoint[]
  cumFlatEquivM: number[]
  cues: TerrainCue[]
  grid: Map<string, number[]>
  /** Meters per degree of longitude at THIS route's latitude (the library
   *  spans Nantucket, New York and Tel Aviv — one global scale is 12% off). */
  kx: number
  // live state
  started: boolean
  idx: number
  progressM: number
  onRouteM: number
  offRouteM: number
  lastLiveDistM: number | null
  /** Started at the route's start while the run was starting — the "same
   *  front door" prior: such a candidate needs less track to lock. */
  startAligned: boolean
}

const GRID_CELL_M = 60
/** SELF-ROUTE: the live track becomes a candidate of its own once this
 *  long — on an out-and-back the outbound leg IS the route for the return
 *  (every trip hill gets crossed twice), on laps lap one is the route for
 *  lap two. No history needed. GUESS — the backtest calibrates. */
export const SELF_ROUTE_MIN_M = 800
/** The newest stretch of the live track is never part of the forward self
 *  candidate — the runner is always "on" his own last meters. */
const SELF_ROUTE_TAIL_M = 200
const SELF_ROUTE_STEP_M = 20
const SELF_ROUTE_REBUILD_M = 100

export class RouteMatcher {
  private readonly cands: Candidate[] = []
  private readonly ky = 110_540
  private current: Candidate | null = null
  private lastFix: Fix | null = null
  private readonly selfRoute: boolean
  private readonly lockMinM: number
  private readonly startPrior: boolean
  /** Downsampled live track (every ~20m) with altitude. */
  private readonly liveTrack: RoutePoint[] = []
  private selfForward: Candidate | null = null
  private selfReversed: Candidate | null = null
  private selfBuiltAtM = 0

  constructor(routes: Route[], opts: { selfRoute?: boolean; lockMinM?: number; startPrior?: boolean } = {}) {
    this.selfRoute = opts.selfRoute ?? true
    this.lockMinM = opts.lockMinM ?? LOCK_MIN_TRACK_M
    this.startPrior = opts.startPrior ?? true
    for (const r of routes) {
      if (r.points.length < 10) continue
      this.cands.push(this.candidate(r, false))
      this.cands.push(this.candidate(r, true))
    }
  }

  /** Grow the live track and (re)build the self candidates. The FORWARD
   *  candidate (laps) keeps its state — its points only ever append. The
   *  REVERSED candidate (out-and-back) is rebuilt only while not started:
   *  once the runner is retracing it, its distances must stay put. */
  private growSelf(fix: Fix): void {
    if (!this.selfRoute) return
    const last = this.liveTrack[this.liveTrack.length - 1]
    if (last && fix.distM - last.distM < SELF_ROUTE_STEP_M) return
    this.liveTrack.push({ lat: fix.lat, lon: fix.lon, distM: fix.distM, altM: fix.altM ?? last?.altM ?? 0 })
    const total = fix.distM
    if (total < SELF_ROUTE_MIN_M + SELF_ROUTE_TAIL_M || total - this.selfBuiltAtM < SELF_ROUTE_REBUILD_M) return
    this.selfBuiltAtM = total
    const body = this.liveTrack.filter((p) => p.distM <= total - SELF_ROUTE_TAIL_M)
    if (body.length < 10) return
    const route: Route = { id: 'self', km: body[body.length - 1].distM / 1000, runs: 0, points: body }
    // Forward: rebuild but carry the live state across (points only grew).
    const fwd = this.candidate(route, false)
    if (this.selfForward) {
      fwd.started = this.selfForward.started
      fwd.idx = this.selfForward.idx
      fwd.progressM = this.selfForward.progressM
      fwd.onRouteM = this.selfForward.onRouteM
      fwd.offRouteM = this.selfForward.offRouteM
      const i = this.cands.indexOf(this.selfForward)
      if (i >= 0) this.cands[i] = fwd
      if (this.current === this.selfForward) this.current = fwd
    } else {
      this.cands.push(fwd)
    }
    this.selfForward = fwd
    // Reversed: frozen once the runner is on it.
    if (!this.selfReversed || !this.selfReversed.started) {
      const rev = this.candidate(route, true)
      if (this.selfReversed) {
        const i = this.cands.indexOf(this.selfReversed)
        if (i >= 0) this.cands[i] = rev
      } else {
        this.cands.push(rev)
      }
      this.selfReversed = rev
    }
  }

  private candidate(route: Route, reversed: boolean): Candidate {
    const total = route.points[route.points.length - 1].distM
    const pts = reversed
      ? [...route.points].reverse().map((p) => ({ ...p, distM: total - p.distM }))
      : route.points
    // Branch probabilities are direction-blind; reversing the points
    // reverses the array. p[i] covers the 100m from point i onward.
    const branch = route.branch && route.branch.length === route.points.length
      ? (reversed ? [...route.branch].reverse() : route.branch)
      : undefined
    const r: Route = { ...route, points: pts, branch }
    const profile: ProfilePoint[] = pts.map((p) => ({ distanceM: p.distM, altitudeM: p.altM }))
    const sm = smoothProfile(profile)
    const cum = new Array<number>(sm.length).fill(0)
    for (let i = 1; i < sm.length; i++) {
      const dd = sm[i].distanceM - sm[i - 1].distanceM
      if (dd <= 0) { cum[i] = cum[i - 1]; continue }
      cum[i] = cum[i - 1] + dd * gapMultiplier((sm[i].altitudeM - sm[i - 1].altitudeM) / dd)
    }
    const kx = 111_320 * Math.cos((pts[0].lat * Math.PI) / 180)
    const grid = new Map<string, number[]>()
    pts.forEach((p, i) => {
      const k = this.cell(kx, p.lat, p.lon)
      const arr = grid.get(k)
      if (arr) arr.push(i)
      else grid.set(k, [i])
    })
    return {
      route: r, reversed, key: `${route.id}${reversed ? ':rev' : ''}`,
      sm, cumFlatEquivM: cum, cues: extractTerrainCues(profile), grid, kx,
      started: false, idx: 0, progressM: 0, onRouteM: 0, offRouteM: 0, lastLiveDistM: null, startAligned: false,
    }
  }

  private cell(kx: number, lat: number, lon: number): string {
    return `${Math.floor((lon * kx) / GRID_CELL_M)},${Math.floor((lat * this.ky) / GRID_CELL_M)}`
  }

  private metersBetween(kx: number, aLat: number, aLon: number, bLat: number, bLon: number): number {
    const dx = (aLon - bLon) * kx
    const dy = (aLat - bLat) * this.ky
    return Math.sqrt(dx * dx + dy * dy)
  }

  /** Nearest route point within the candidate's forward window (or the
   *  whole route via the grid when not started) and the runner's CONTINUOUS
   *  progress: the fix projected onto the segment through that point. Route
   *  points are 20m apart — a point-snapped progress moved in 20m jumps, and
   *  the fresh-cut window (4s ≈ 12m) fell between two jumps every time.
   *  Global search is DIRECTION-AWARE: on an out-and-back the outbound and
   *  return legs overlap within meters — the runner's own heading picks the
   *  leg (a leg-blind start put the pointer on the wrong leg, stalled it,
   *  and churned the lock at every turnaround).
   *  Returns [index, residualM, progressM]. */
  private nearest(c: Candidate, fix: Fix, heading: [number, number] | null): [number, number, number] {
    const pts = c.route.points
    let best = -1
    let bestD = Infinity
    if (c.started) {
      const from = c.progressM - BACK_WINDOW_M
      const to = c.progressM + FORWARD_WINDOW_M
      for (let i = Math.max(0, c.idx - 3); i < pts.length && pts[i].distM <= to; i++) {
        if (pts[i].distM < from) continue
        const d = this.metersBetween(c.kx, fix.lat, fix.lon, pts[i].lat, pts[i].lon)
        if (d < bestD) { bestD = d; best = i }
      }
    } else {
      const cx = Math.floor((fix.lon * c.kx) / GRID_CELL_M)
      const cy = Math.floor((fix.lat * this.ky) / GRID_CELL_M)
      for (let dx = -1; dx <= 1; dx++) {
        for (let dy = -1; dy <= 1; dy++) {
          const arr = c.grid.get(`${cx + dx},${cy + dy}`)
          if (!arr) continue
          for (const i of arr) {
            const d = this.metersBetween(c.kx, fix.lat, fix.lon, pts[i].lat, pts[i].lon)
            if (d > NEAR_M || d >= bestD) continue
            if (heading && i + 1 < pts.length) {
              const rx = (pts[i + 1].lon - pts[i].lon) * c.kx
              const ry = (pts[i + 1].lat - pts[i].lat) * this.ky
              if (rx * heading[0] + ry * heading[1] <= 0) continue // wrong way along this leg
            }
            bestD = d
            best = i
          }
        }
      }
    }
    if (best < 0) return [-1, Infinity, c.progressM]
    // Project onto the better of the two segments touching the nearest point.
    let progress = pts[best].distM
    let residual = bestD
    for (const [a, b] of [[best - 1, best], [best, best + 1]] as const) {
      if (a < 0 || b >= pts.length) continue
      const ax = pts[a].lon * c.kx, ay = pts[a].lat * this.ky
      const bx = pts[b].lon * c.kx, by = pts[b].lat * this.ky
      const px = fix.lon * c.kx, py = fix.lat * this.ky
      const vx = bx - ax, vy = by - ay
      const len2 = vx * vx + vy * vy
      if (len2 <= 0) continue
      const tt = Math.max(0, Math.min(1, ((px - ax) * vx + (py - ay) * vy) / len2))
      const qx = ax + tt * vx, qy = ay + tt * vy
      const d = Math.hypot(px - qx, py - qy)
      if (d < residual) {
        residual = d
        progress = pts[a].distM + tt * (pts[b].distM - pts[a].distM)
      }
    }
    return [best, residual, progress]
  }

  /** Feed one GPS fix. Call at whatever rate fixes arrive (≈1Hz). */
  update(fix: Fix): void {
    const dLive = this.lastFix ? Math.max(0, fix.distM - this.lastFix.distM) : 0
    // Runner's heading from the last two fixes (meters), only once he has moved.
    let heading: [number, number] | null = null
    if (this.lastFix) {
      const kx = 111_320 * Math.cos((fix.lat * Math.PI) / 180)
      const hx = (fix.lon - this.lastFix.lon) * kx
      const hy = (fix.lat - this.lastFix.lat) * this.ky
      if (Math.hypot(hx, hy) >= 3) heading = [hx, hy]
    }
    this.lastFix = fix
    this.growSelf(fix)
    for (const c of this.cands) {
      const [i, res, prog] = this.nearest(c, fix, heading)
      if (!c.started) {
        if (i >= 0 && res <= NEAR_M) {
          c.started = true
          c.idx = i
          c.progressM = prog
          c.onRouteM = 0
          c.offRouteM = 0
          c.startAligned = this.startPrior && prog < 100 && fix.distM < 100 && c.route.id !== 'self'
        }
        continue
      }
      if (i >= 0 && res <= NEAR_M) {
        // Progress must move forward: a runner standing still or looping
        // back doesn't advance the pointer past where he is.
        if (prog >= c.progressM - BACK_WINDOW_M) {
          c.idx = i
          c.progressM = Math.max(c.progressM, prog)
        }
        c.onRouteM += dLive
        c.offRouteM = 0
      } else if (i < 0 || res > LOST_M) {
        c.offRouteM += dLive
        if (c.offRouteM >= LOST_TRAVEL_M) {
          c.started = false
          c.onRouteM = 0
          c.offRouteM = 0
          if (this.current === c) this.current = null
        }
      }
      // NEAR_M < res ≤ LOST_M: ambiguous (wide road, GPS wobble) — hold.
    }
    // Lock policy: the candidate with the most on-route travel past the
    // threshold wins; habit (runs) then length break ties. The current lock
    // is sticky — a rival must beat it clearly, not merely tie it.
    let best: Candidate | null = null
    for (const c of this.cands) {
      if (!this.lockEligible(c)) continue
      if (!best || this.better(c, best)) best = c
    }
    if (this.current && this.lockEligible(this.current)) {
      if (best && best !== this.current && best.onRouteM > this.current.onRouteM + this.lockMinM) this.current = best
    } else {
      this.current = best
    }
  }

  private lockEligible(c: Candidate): boolean {
    if (!c.started) return false
    return c.onRouteM * (c.startAligned ? 1.5 : 1) >= this.lockMinM
  }

  private better(a: Candidate, b: Candidate): boolean {
    if (a.onRouteM !== b.onRouteM) return a.onRouteM > b.onRouteM
    if (a.route.runs !== b.route.runs) return a.route.runs > b.route.runs
    return a.route.km > b.route.km
  }

  /** The locked route (id, direction, progress), or null. `agreement` is
   *  the share of lock-eligible candidates that put the runner at the same
   *  place 300m ahead — 1.0 means every plausible route agrees on the
   *  future; less means a fork is coming that history has taken both ways. */
  get lock(): { routeId: string; reversed: boolean; progressM: number; remainingM: number; agreement: number } | null {
    const c = this.current
    if (!c) return null
    const total = c.route.points[c.route.points.length - 1].distM
    return { routeId: c.route.id, reversed: c.reversed, progressM: c.progressM, remainingM: total - c.progressM, agreement: this.agreement() }
  }

  /** Probability (from history) that the runner is still on the locked
   *  route `aheadM` from here — the product of the per-100m continuation
   *  shares along the way. What a "final kilometer" claim should check. */
  pAhead(aheadM: number): number {
    const c = this.current
    if (!c) return 0
    return this.pReach(c, c.progressM + aheadM)
  }

  private pReach(c: Candidate, toRouteDistM: number): number {
    const b = c.route.branch
    if (!b) return 1
    const pts = c.route.points
    let p = 1
    for (let i = c.idx; i < pts.length && pts[i].distM < toRouteDistM; i += 5) p *= b[i] ?? 1
    return p
  }

  /** Where candidate c says the runner will be `aheadM` past its progress. */
  private pointAhead(c: Candidate, aheadM: number): RoutePoint | null {
    const target = c.progressM + aheadM
    const pts = c.route.points
    for (let i = c.idx; i < pts.length; i++) if (pts[i].distM >= target) return pts[i]
    return null
  }

  private agreement(aheadM = 300): number {
    const cur = this.current
    if (!cur) return 0
    const mine = this.pointAhead(cur, aheadM)
    if (!mine) return 1
    let eligible = 0
    let agree = 0
    for (const c of this.cands) {
      if (!this.lockEligible(c)) continue
      eligible++
      const p = this.pointAhead(c, aheadM)
      if (p && this.metersBetween(cur.kx, mine.lat, mine.lon, p.lat, p.lon) <= 40) agree++
    }
    return eligible ? agree / eligible : 1
  }

  /** Terrain cues AHEAD on the locked route, projected onto the live odometer. */
  aheadCues(): AheadCue[] {
    const c = this.current
    const fix = this.lastFix
    if (!c || !fix) return []
    const here = this.flatEquivAt(c, c.progressM)
    // A contested future halves the cue's confidence: below the 0.5 gate it
    // stays reactive-only (design doc's TerrainCue contract).
    const agreement = this.agreement()
    const out: AheadCue[] = []
    for (const cue of c.cues) {
      if (cue.distanceM <= c.progressM) continue
      const pReach = this.pReach(c, cue.distanceM)
      out.push({
        ...cue,
        pReach,
        confidence: cue.confidence * (agreement >= 1 ? 1 : 0.5) * pReach,
        liveDistanceM: fix.distM + (cue.distanceM - c.progressM),
        flatEquivRemainingM: this.flatEquivAt(c, cue.distanceM) - here,
        key: `${c.key}@${Math.round(cue.distanceM)}`,
      })
    }
    return out
  }

  /** The route's elevation ahead of the runner in LIVE-odometer space —
   *  what a "route ahead" panel or the coaching layer reads. */
  aheadProfile(maxM = 2000): ProfilePoint[] {
    const c = this.current
    const fix = this.lastFix
    if (!c || !fix) return []
    const out: ProfilePoint[] = []
    for (const p of c.sm) {
      if (p.distanceM < c.progressM) continue
      if (p.distanceM > c.progressM + maxM) break
      out.push({ distanceM: fix.distM + (p.distanceM - c.progressM), altitudeM: p.altitudeM })
    }
    return out
  }

  private flatEquivAt(c: Candidate, d: number): number {
    const sm = c.sm
    const cum = c.cumFlatEquivM
    let lo = 0
    let hi = sm.length - 1
    while (lo < hi) {
      const mid = (lo + hi) >> 1
      if (sm[mid].distanceM < d) lo = mid + 1
      else hi = mid
    }
    if (lo === 0) return 0
    const a = sm[lo - 1]
    const b = sm[lo]
    const f = Math.min(1, Math.max(0, (d - a.distanceM) / Math.max(1, b.distanceM - a.distanceM)))
    return cum[lo - 1] + f * (cum[lo] - cum[lo - 1])
  }
}
